In [13]:
import numpy as np

In [ ]:
def compute_vector_length(vector):
    return np.sqrt(np.sum(vector**2))

In [ ]:
vector = np.array ([-2 , 4 , 9 , 21])
result = compute_vector_length(vector)
print(round(result,2))

23.28


In [ ]:
def compute_dot_product(vector1, vector2):
    return np.sum(vector1 * vector2)

In [ ]:
v1 = np.array([0 , 1 , -1 , 2])
v2 = np.array([2 , 5 , 1 , 0])
result = compute_dot_product(v1 , v2)
print(round(result ,2))

4


In [ ]:
x = np.array([[1 , 2],[3 , 4]])
k = np.array([1 , 2])
print ('result \n', x.dot (k) )

result 
 [ 5 11]


In [ ]:
x = np. array([[ -1 , 2] ,[3 , -4]])
k = np. array([1 , 2])
print('result \n', x@k )

result 
 [ 3 -5]


In [ ]:
def matrix_multi_vector(matrix, vector):
    return np.dot(matrix, vector)

In [ ]:
m = np.array([[ -1 , 1 , 1] , [0 , -4 , 9]])
v = np.array([0 , 2 , 1])
result = matrix_multi_vector(m, v)
print(result)

[3 1]


In [ ]:
def matrix_multi_matrix(matrix1, matrix2):
    return np.dot(matrix1, matrix2)

In [ ]:
m1 = np.array([[0 , 1 , 2] , [2 , -3 , 1]])
m2 = np.array([[1 , -3] ,[6 , 1] , [0 , -1]])
result = matrix_multi_matrix(m1 , m2)
print(result)

[[  6  -1]
 [-16 -10]]


In [ ]:
m1 = np.eye(3)
m2 = np.array([[1 , 1 , 1] ,[2 , 2 , 2] , [3 , 3 , 3]])
result = m1@m2
print(result)


[[1. 1. 1.]
 [2. 2. 2.]
 [3. 3. 3.]]


In [5]:
m1 = np.eye(2)
print(m1)
m1 = np.reshape(m1,(-1 ,4))
print(m1)
m2 = np.array([[1 , 1 , 1 , 1] ,[2 , 2 , 2 , 2] , [3 , 3 , 3 , 3] , [4 , 4 , 4 , 4]])
result = m1@m2
print(result)

[[1. 0.]
 [0. 1.]]
[[1. 0. 0. 1.]]
[[5. 5. 5. 5.]]


In [6]:
m1 = np. array([[1 , 2] , [3 , 4]])
m1 = np. reshape (m1 ,( -1 ,4), "F") [0]
m2 = np. array([[1 , 1 , 1 , 1],[2 , 2 , 2 , 2],[3 , 3 , 3 , 3],[4 , 4 , 4 , 4]])
result = m1@m2
print ( result )

[29 29 29 29]


In [8]:
def inverse_matrix(matrix):
    return np.linalg.inv(matrix)

## Inverse Matrix

*   Matrix A:
\begin{equation}
A = \begin{bmatrix} -2 & 6 \\ 8 & -4 \end{bmatrix}
\end{equation}

*   Determinant of A:
\begin{equation}
det(A) = (-2)*(-4) - 8*6 = -40
\end{equation}

*   Inverse matrix of A:
\begin{equation}
  A^{-1} = \frac{1}{-40}\begin{bmatrix} -4 & -6 \\ -8 & -2 \end{bmatrix} = \begin{bmatrix} 0.1 & 0.15 \\ 0.2 & 0.05 \end{bmatrix}
\end{equation}



In [9]:
matrix = np.array([[-2, 6], [8, -4]])
result = inverse_matrix(matrix)
print(result)

[[0.1  0.15]
 [0.2  0.05]]


In [11]:
def compute_eigenvalues_eigenvectors(matrix):
    eigenvalues, eigenvectors = np.linalg.eig(matrix)
    return eigenvalues, eigenvectors

In [18]:
matrix = np.array ([[0.9 , 0.2] , [0.1 , 0.8]])
eigenvalues, eigenvectors = compute_eigenvalues_eigenvectors( matrix )
print(eigenvalues)
print(eigenvectors)

[1.  0.7]
[[ 0.89442719 -0.70710678]
 [ 0.4472136   0.70710678]]


## Eigenvector and Eigenvalue of Matrix A

*   Given the matrix A:
\begin{equation}
A = \begin{bmatrix}
0.9 & 0.2 \\
0.1 & 0.8
\end{bmatrix}
\end{equation}

*   Compute the Eigenvalues

The eigenvalues $\lambda$ are found by solving the characteristic equation:
\begin{equation}
\text{det}(A - \lambda I) = 0
\\
A - \lambda I = \begin{bmatrix}
0.9 - \lambda & 0.2 \\
0.1 & 0.8 - \lambda
\end{bmatrix}
\\
⇒ \text{det}(A - \lambda I) = (0.9 - \lambda)(0.8 - \lambda) - (0.2)(0.1) = \lambda^2 - 1.7\lambda + 0.7 = 0
\end{equation}

Solving the quadratic equation:
\begin{equation}
\lambda = \frac{1.7 \pm \sqrt{(1.7)^2 - 4 \cdot 0.7}}{2} = \frac{1.7 \pm \sqrt{2.89 - 2.8}}{2} = \frac{1.7 \pm \sqrt{0.09}}{2} = \frac{1.7 \pm 0.3}{2}
\end{equation}

Thus, the eigenvalues are:
\begin{equation}
⇒ \lambda_1 = 1, \quad \lambda_2 = 0.7
\end{equation}

*  Compute the Eigenvectors

For $\lambda_1 = 1$
\begin{equation}
(A - I)v = 0
\\
⇒ \begin{bmatrix}
0.9 - 1 & 0.2 \\
0.1 & 0.8 - 1
\end{bmatrix} \begin{bmatrix}
v_1 \\
v_2
\end{bmatrix} = 0
\\
⇒ \begin{bmatrix}
-0.1 & 0.2 \\
0.1 & -0.2
\end{bmatrix} \begin{bmatrix}
v_1 \\
v_2
\end{bmatrix} = 0
\end{equation}

This reduces to:
\begin{equation}
-0.1v_1 + 0.2v_2 = 0
\\
⇒ v_2 = \frac{v_1}{2}
\end{equation}

Thus, the eigenvector corresponding to $\lambda_1 = 1$ is:
\begin{equation}
v_1 = \begin{bmatrix}
1 \\ 0.5
\end{bmatrix}
\end{equation}

Normalize $v_1$:
\begin{equation}
||v_1|| = \sqrt{1^2 + 0.5^2} = \sqrt{1.25} \approx 1.118
\\
⇒ v_1 = \begin{bmatrix}
\frac{1}{1.118} \\
\frac{0.5}{1.118}
\end{bmatrix} = \begin{bmatrix}
\frac{1}{\sqrt{1.25}} \\
\frac{0.5}{\sqrt{1.25}}
\end{bmatrix} \approx \begin{bmatrix}
0.894 \\
0.447
\end{bmatrix}
\end{equation}

For $ \lambda_2 = 0.7 $:
\begin{equation}
(A - 0.7I)v = 0
\\
⇒ \begin{bmatrix}
0.9 - 0.7 & 0.2 \\ 0.1 & 0.8 - 0.7
\end{bmatrix} \begin{bmatrix}
v_1 \\ v_2
\end{bmatrix} = 0
\\
⇒ \begin{bmatrix}
0.2 & 0.2 \\ 0.1 & 0.1
\end{bmatrix}
\begin{bmatrix}
v_1 \\ v_2
\end{bmatrix} = 0
\end{equation}

This reduces to:
\begin{equation}
0.2v_1 + 0.2v_2 = 0
\\
⇒ v_1 = -v_2
\end{equation}

Thus, the eigenvector corresponding to $ \lambda_2 = 0.7 $ is:
\begin{equation}
v_2 = \begin{bmatrix}
1 \\ -1
\end{bmatrix}
\end{equation}

Normalize $v_2 $:
\begin{equation}
||v_2|| = \sqrt{1^2 + (-1)^2} = \sqrt{2}
\\
⇒ v_2 = \begin{bmatrix}
\frac{1}{\sqrt{2}} \\
\frac{-1}{\sqrt{2}}
\end{bmatrix} \approx \begin{bmatrix}
0.707 \\
-0.707
\end{bmatrix}
\end{equation}

## Summary
The normalized eigenvectors and eigenvalues are:
\begin{equation}
\lambda_1 = 1, \quad v_1 = \begin{bmatrix}
0.894 \\
0.447
\end{bmatrix}
\\
\lambda_2 = 0.7, \quad v_2 = \begin{bmatrix}
0.707 \\
-0.707
\end{bmatrix}
\end{equation}


In [15]:
def compute_cosine(vector1, vector2):
    return np.dot(vector1, vector2) / (np.linalg.norm(vector1) * np.linalg.norm(vector2))

## Cosine Similarity

*   Vector X and Vector Y:
\begin{equation}
x = \begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix}
\\
y = \begin{bmatrix} 1 \\ 0 \\ 3 \\ 0 \end{bmatrix}
\end{equation}

*   Cosine Simularity
\begin{equation}
x.y = y = 1*1 + 3*3 = 10
\\
||x|| = \sqrt(1^2 + 2^2 + 3^2 + 4^2) = \sqrt30
\\
||y|| = \sqrt(1^2 + 0^2 + 3^2 + 0^2) = \sqrt10
\\
\Rightarrow cs(x,y) = \frac{x.y}{||x||.||y||} = \frac{1}{\sqrt3}
\end{equation}


In [16]:
x = np.array([1 , 2 , 3 , 4])
y = np.array([1 , 0 , 3 , 0])
result = compute_cosine(x,y)
print(round(result,3))

0.577
